File: ST_engine/src/main/entity.cpp
#include "entity.hpp"

//default constructor
entity::entity(unsigned int id){
    ID = id;
}

//GENERAL FUNCTIONS//
Uint64 entity::get_ID(){
    return ID;
}

void entity::set_active(bool arg){
	active = arg;
}

bool entity::is_active(){
	return active;
}

int entity::get_x(){
    return x;
}

int entity::get_y(){
    return y;
}

void entity::set_x(int X){
    x = X;
    update_collision_box(x, y);
}

void entity::set_y(int Y){
    y = Y;
    update_collision_box(x, y);
}

void entity::set_static(bool arg){
    static_ = arg;
}

bool entity::is_static(){
    return static_;
}

void entity::set_velocity_x(int arg){
    velocity_x = arg;
}

void entity::set_velocity_y(int arg){
    velocity_y = arg;
}

int entity::get_velocity_x(){
    return velocity_x;
}

int entity::get_velocity_y(){
    return velocity_y;
}

//TEXTURE FUNCTIONS//

int entity::get_tex_w(){
    return tex_w;
}

int entity::get_tex_h(){
    return tex_h;
}

void entity::set_tex_w(int width){
    tex_w = width;
}

void entity::set_tex_h(int height){
    tex_h = height;
}

size_t entity::get_texture(){
    return texture;
}

void entity::set_visible(bool arg){
    visibility = arg;
}

void entity::set_texture(std::string arg){
    std::hash<std::string> hash_f;
    texture = hash_f(arg);
}

bool entity::is_visible(){
    return visibility;
}

//PHYSICS FUNTIONS//

int entity::get_col_x(){
    return col_x;
}

int entity::get_col_y(){
    return col_y;
}

int entity::get_col_x_offset(){
	return offset_x;
}

int entity::get_col_y_offset(){
	return offset_y;
}

void entity::set_collision_box(int offsetX, int offsetY, int X, int Y){
    if(X != 0 && Y != 0){
        col_x = X - 1;
        col_y = -Y + 1;
        offset_x = offsetX;
        offset_y = offsetY;
    }else{
        col_x = 0;
        col_y = 0;
        set_affected_by_physics(false);
    }
    update_collision_box(x, y);
}

void entity::update_collision_box(int x, int y){
    collision.left = x + offset_x;
    collision.right = x + col_x + offset_x;
    collision.top = y + col_y + offset_y;
    collision.bottom = y + offset_y;
}

void entity::set_affected_by_physics(bool arg){
    affected_by_physics = arg;
}

bool entity::is_affected_by_physics(){
    return affected_by_physics;
}

bool entity::collides(entity other){
    return !((collision.bottom < other.collision.top) | (collision.top > other.collision.bottom)
    | (collision.left > other.collision.right) | (collision.right < other.collision.left));
}

void entity::set_mass(int arg){
    mass = arg;
}

int entity::get_mass(){
    return mass;
}

//ANIMATION FUNCTIONS//

void entity::set_animation(int arg){
    animation = arg;
}

int entity::get_animation(){
    return animation;
}

void entity::set_animation_num(int arg){
    animation_num = arg;
}

int entity::get_animation_num(){
    return animation_num;
}

void entity::set_sprite_num(int arg){
    sprite_num = arg;
}

int entity::get_sprite_num(){
    return sprite_num;
}

File: ST_engine/gamedata/lua/objects/entity.lua
-- This file is part of the "ST" project.
-- You may use, distribute or modify this code under the terms
-- of the GNU General Public License version 2.
-- See LICENCE.txt in the root directory of the project.
--
-- Author: Maxim Atanasov
-- E-mail: maxim.atanasov@protonmail.com

--tells if the mouse is over an objects texture
function mouseOverTexture(object)
    return mouseOverTextureLua(object.ID)
end

--tells if the mouse is over an objects collisions box
function mouseOver(object)
    return mouseOverLua(object.ID)
end

function newEntityID()
    local temp = currentID
    currentID = currentID + 1
    return temp
end

--default constructor for all entities
function newEntity(self, x, y)
    local o = {}
    setmetatable(o, self)
    self.__index = self
    if(x ~= nil and y ~= nil) then
        --general
        o.ID = newEntityID()
	    createEntity()
        setEntityX(o.ID, x)
        setEntityY(o.ID, y)

        --texture
        setEntityTexture(o.ID, o.texture)
        setEntityTexW(o.ID, o.texWidth)
        setEntityTexH(o.ID, o.texHeight)
        setEntityVisible(o.ID, o.isVisible)

        --animation
        setEntitySpriteNum(o.ID, o.spriteNum)
        setEntityAnimationNum(o.ID, o.animationNum)

        --physics
        setEntityAffectedByPhysics(o.ID, o.affectedByPhysics)
	    setEntityCollisionBox(o.ID, o.offsetColX, o.offsetColY, o.colX, o.colY)
    end
    return o
end

--generic game entity, from which all other entities inherit
entity = {
	texture = "";
    offsetColX = 0;
    offsetColY = 0;
	colX = 0;
	colY = 0;
    isVisible = false;
    affectedByPhysics = false;
    mass = 0;

    --animations
    spriteNum = 1;
    animationNum = 1;
}

--create a new instance of an entity (calls the constructor above)
function entity:new(x, y)
    return newEntity(self, x, y)
end

function entity:setAffectedByPhysics(arg)
    self.affectedByPhysics = arg
    setEntityAffectedByPhysics(self.ID, arg)
end

--set the x position of an entity
function entity:setX(x)
    setEntityX(self.ID, x)
end

--set the y position of an entity
function entity:setY(y)
    setEntityY(self.ID, y)
end

--set the x and y position of an entity
function entity:setXY(x, y)
	setEntityX(self.ID, x)
	setEntityY(self.ID, y)
end

--get the x coordinate of an entity
function entity:getX()
    return getEntityX(self.ID)
end

--get the y coordinate of an entity
function entity:getY()
    return getEntityY(self.ID)
end

--tells if an entity (without hard collision) is over another entity
function entity:overObject(other)
    return entityCollides(self.ID, other.ID)
end

--set the Texture width of an entity
function entity:setTexW(arg)
    setEntityTexW(self.ID, arg)
end

--set the Texture height of an entity
function entity:setTexH(arg)
    setEntityTexH(self.ID, arg)
end

function entity:setTextureScale(x, y)
    self.textureScaleX = x;
    self.textureScaleY = y;
    setEntityTextureScale(self.ID, x, y)
end

--get the Texture width of an entity
function entity:getTexW()
    return getEntityTexW(self.ID)
end

--get the Texture height of an entity
function entity:getTexH()
    return getEntityTexH(self.ID)
end

--set the horizontal velocity of an entity
function entity:setVelocityX(arg)
    setEntityVelocityX(self.ID, arg)
end

--set the vertical velocity of an entity
function entity:setVelocityY(arg)
    setEntityVelocityY(self.ID, arg)
end

--get the horizontal velocity of an entity
function entity:getVelocityX()
    return getEntityVelocityX(self.ID)
end

--get the vertical velocity of an entity
function entity:getVelocityY()
    return getEntityVelocityY(self.ID)
end

--set the collision box of an entity
function entity:setCollision(offsetX, offsetY, width, height)
    self.colX = width
    self.colY = height
    self.offsetColX = offsetX;
    self.offsetColY = offsetY;
    setEntityCollisionBox(self.ID, offsetX, offsetY, width, height)
end

--scale the collision box of an entity`
function entity:setCollisionScale(x, y)
    setEntityCollisionBox(self.ID, 0, 0, x*self.colX, y*self.colY)
end

--get the Width of the collision box of an entity
function entity:getColX()
    return getEntityColX(self.ID)
end

--get the Height of the collision box of an entity
function entity:getColY()
    return getEntityColY(self.ID)
end

--set the Visibility of an entity
function entity:setVisible(arg)
    self.isVisible = arg;
    setEntityVisible(self.ID, arg)
end

--set weather an entity is static (always drawn at the same coordinates, regardless of camera position)
function entity:setStatic(arg)
    setEntityStatic(self.ID, arg)
end

--set the Texture of an entity (filepath)
function entity:setTexture(arg)
    self.texture = arg
    setEntityTexture(self.ID, arg)
end

--play the corresponding animation of an entity (the corresponding index in the spritesheet)
function entity:playAnimation(arg)
    setEntityAnimation(self.ID, arg)
end

--delete entity
function entity:delete()
    setEntityVisible(self.ID, false)
    self:setCollision(0, 0, 0, 0)
    setEntityActive(self.ID, false)
    self = nil
end
